### PR TITLE
Initial type definition generation tools

### PR DESCRIPTION
**Pull request by colelawrence**
In reference to #43, this is the initial approach I have in mind for creating typing files.

The bulk of the logic involved here reside in the [`./definitions/transpiler.ts`](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/definitions/transpiler.ts) and [`./definitions/types.template.ts`](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/definitions/types.template.ts).

The [transpiler](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/definitions/transpiler.ts) transpiles macro block in order to simplify the complexity of variable argument functions and constructors. These macros are explicitly defined in the transpiler source and enable things like the following:

[**Source**](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/definitions/types.template.ts)
```typescript
    // In-progress: All encompassing objects
    export type TapObject<
        T = any,
        // Loop produces A = any, B = any, C = any, etc... using /\bA\b/g replace
        // >>> loopLetter
        A = any,
        // <<< loopLetter
    > =
        // >>> loopTapGenerics
        | TapObject1<A> | TapCtxObject1<T, A>
        // <<< loopTapGenerics
    ;
```
[**Build**](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/types.build.d.ts)
```typescript
    // In-progress: All encompassing objects
    export type TapObject<
        T = any,
        // Loop produces A = any, B = any, C = any, etc... using /\bA\b/g replace
        A = any,
        B = any,
        C = any,
        D = any,
    > =
        | TapObject0 | TapCtxObject0<T>

        | TapObject1<A> | TapCtxObject1<T, A>

        | TapObject2<A, B> | TapCtxObject2<T, A, B>

        | TapObject3<A, B, C> | TapCtxObject3<T, A, B, C>

        | TapObject4<A, B, C, D> | TapCtxObject4<T, A, B, C, D>
    ;
```

Please add thoughts and comments for discussion.

### CODE REVIEW

File: definitions/package.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "definitions",
+  "version": "1.0.0",
+  "scripts": {
+    "test": "ts-node ./build.ts"
+  },
+  "dependencies": {
+    "tapable": "../"
```
Comment (colelawrence) [inline definitions/package.json]: This is not yet used... I'm not exactly sure how I would like to actually test the type definitions for comprehensive coverage.

**Comment by colelawrence** [comment]
I think the snapshots may be out of date. I can do a brand new generation off master and the updated jest snapshot definitions are essentially all the same except the new snapshots use `this._x1 = someval` as opposed to `_x1 = someval`

Is this just happening on my pull?

**Comment by colelawrence** [comment]
I have also created an example usage file as [test-types.ts](https://github.com/colelawrence/tapable/blob/feature/cl-12-23-type-definitions/definitions/test-types.ts) which may not be very idiomatic usage of Tapable right now, but does seem to produce correct output.

File: definitions/package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "definitions",
+  "version": "1.0.0",
+  "scripts": {
+    "build": "ts-node ./build.ts",
+    "test": "ts-node ./test-types.ts"
+  },
+  "dependencies": {
+    "tapable": "../"
```
Comment (colelawrence) [inline definitions/package.json]: I think this uses the typings provided by `package.json: "typings": "./types.build.d.ts"`

File: definitions/types.template.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
+// >>> replace 'module tapable', 'declare module "tapable"'
```
Comment (colelawrence) [inline definitions/types.template.ts]: Notice use of parameters for calling replace code generation

**Comment by TheLarkInn** [comment]
Hey sorry I missed this PR thus late response. Both @sokra and myself will prioritize this. Thank you for the PR!!!!!! 🙇🙇🙇 @danielrosenwasser would you mind reviewing also?

**Comment by colelawrence** [comment]
No problem, Sean. I could use help on the accuracy of everything. As far as
I can tell it all checks out and can catch very subtle type compatibility
errors for the tests I've been doing. I mostly need help understanding the
answers to the aforementioned questions in #43 🤔

**Comment by DanielRosenwasser** [comment]
Hey @colelawrence, I think this PR is highly appreciated. However, our team strongly recommends keeping these declarations (or updating them) on DefinitelyTyped so that the types-publisher can be responsible for updating with new versions. Additionally, keeping types on DefinitelyTyped means that the ecosystem can be tested more holistically and reviewed by the team.

Otherwise, @sokra and @TheLarkInn will have to maintain these files and publish new versions every update. There's also a level of blind trust there, so we've found that having library authors maintain type declarations within their own repo is an unfair burden on the maintainers.

**Comment by TheLarkInn** [comment]
That is fine by me. Cole do you have any objections to this?

**Comment by colelawrence** [comment]
I don't have any problems with this approach at all.

It sounds like a really great way to distribute these types. Thanks for the
suggestion, @DanielRosenwasser

I would like to open a PR over in the DefinitelyTyped repo and get a quick
review that these types look correct. I would still like some guidance on
the typings questions I prompted in #43 and I'd like to know what version
of tapable these types should be tracked to / should be released with.

I think these types correspond with the *wip* branch which may be versioned
as 1.0.0-beta.4

So, in order to make these types useable, and not confused with types for
the current stable version of tapable, do we need to do anything special
Daniel or Sean?

**Comment by colelawrence** [comment]
I think it would be safe to close this in favor of opening these changes up as a @types PR, then we can resume discussion there if there are any questions or bugfixes.

**Comment by TheLarkInn** [comment]
Thanks a lot Cole and sorry for any of the confusion. Is there a PR I can
track? We've really appreciated the exploration on this.

**Comment by colelawrence** [comment]
@TheLarkInn I left my progress in the branch I was working against at: https://github.com/colelawrence/tapable/tree/feature/cl-12-23-type-definitions

**Comment by colelawrence** [comment]
I also want to point out that I might have hit andead end with this approach, as explained in #43

@thelarkinn

**Pull request closed without merging**